pkg/util/watch/metrics.go

```diff
@@ -0,0 +1,180 @@
+	// metrics collected for the Watch call
+	Instance string
+}
+
+const metricInstanceLabel = "watcher_instance"
+
+// NewMetrics creates a new set of metrics for one or many Watch calls
+//
+// All metrics' names will be prefixed with the provided string.
+func NewMetrics(prefix string) Metrics {
+	return Metrics{
+		clientCallsTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: fmt.Sprint(prefix, "client_calls_total"),
+				Help: "Number of calls to k8s client.{Watch,List}, labeled by method",
+			},
+			[]string{metricInstanceLabel, "method"},
+		),
+		relistRequestsTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: fmt.Sprint(prefix, "relist_requests_total"),
+				Help: "Number of internal manual relisting requests",
+			},
+			[]string{metricInstanceLabel},
+		),
+		eventsTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: fmt.Sprint(prefix, "events_total"),
+				Help: "Number of k8s watch.Events that have occurred, including errors, labeled by type",
+			},
+			[]string{metricInstanceLabel, "type"},
+		),
+		errorsTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: fmt.Sprint(prefix, "errors_total"),
+				Help: "Number of errors, either error events or re-list errors, labeled by source",
+			},
+			[]string{metricInstanceLabel, "source"},
+		),
+		aliveCurrent: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: fmt.Sprint(prefix, "alive_current"),
+				Help: "For each watcher, 1 iff the watcher is currently running or failing, else 0",
+			},
+			[]string{metricInstanceLabel},
+		),
+		failingCurrent: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: fmt.Sprint(prefix, "failing_current"),
+				Help: "For each watcher, 1 iff the watcher's last request failed *and* it's waiting to retry, else 0",
+			},
+			[]string{metricInstanceLabel},
+		),
+	}
+}
+
+// MustRegister registers all the collectors in the Metrics
+func (m *Metrics) MustRegister(reg *prometheus.Registry) {
+	reg.MustRegister(m.clientCallsTotal)
+	reg.MustRegister(m.relistRequestsTotal)
+	reg.MustRegister(m.eventsTotal)
+	reg.MustRegister(m.errorsTotal)
+	reg.MustRegister(m.aliveCurrent)
+	reg.MustRegister(m.failingCurrent)
+}
+
+///////////////////////////////////////////////
+// Internal helper methods for MetricsConfig //
+///////////////////////////////////////////////
+
+func (m *MetricsConfig) alive() {
+	if m != nil {
+		m.aliveCurrent.WithLabelValues(m.Instance).Inc()
+	}
+}
+
+func (m *MetricsConfig) unalive() {
+	if m != nil {
+		m.aliveCurrent.WithLabelValues(m.Instance).Dec()
+	}
+}
+
+func (m *MetricsConfig) failing(store *bool) {
+	var wasFailing bool
+	wasFailing, *store = *store, true
+
+	if m != nil && !wasFailing {
+		m.failingCurrent.WithLabelValues(m.Instance).Inc()
+	}
+}
+
+func (m *MetricsConfig) unfailing(store *bool) {
+	var wasFailing bool
+	wasFailing, *store = *store, false
```

Comment:
I guess I'd probably store (an `atomic.Bool` (??)) in the `MetricsConfig` struct and track the state of failing/not-failing there rather than passing a pointer in/out and mutating here?